rust/core-lib/src/tabs.rs

```diff
@@ -289,38 +308,36 @@ impl<W: Write + Send + 'static> Documents<W> {
         json!(view_id)
     }
 
-    fn do_close_view(&mut self, view_id: &ViewIdentifier) {
-        self.close_view(view_id);
+    fn close_view(&mut self, view_id: &ViewIdentifier) {
+        self.buffers.close_view(view_id);
     }
 
-    fn new_empty_view(&mut self, rpc_peer: &MainPeer<W>,
-                      view_id: &str, buffer_id: BufferIdentifier) {
+    fn new_empty_view(&mut self, rpc_peer: &MainPeer<W>, view_id: &ViewIdentifier,
+                      buffer_id: BufferIdentifier) {
         let editor = Editor::new(self.new_tab_ctx(rpc_peer), view_id);
-        let mut inner = self.buffers.lock();
-        inner.views.insert(view_id.to_owned(), buffer_id.clone());
-        inner.editors.insert(buffer_id, editor);
+        self.buffers.add_editor(view_id, &buffer_id, editor, None);
     }
 
     fn new_view_with_file(&mut self, rpc_peer: &MainPeer<W>, view_id: &ViewIdentifier,
                           buffer_id: BufferIdentifier, path: &Path) {
         match self.read_file(&path) {
             Ok(contents) => {
-                let editor = Editor::with_text(self.new_tab_ctx(rpc_peer), view_id, contents);
-                {
-                    let mut inner = self.buffers.lock();
-                    inner.views.insert(view_id.to_owned(), buffer_id.clone());
-                    inner.editors.insert(buffer_id, editor);
-                }
-                self.buffers.add_file(path, view_id);
+                let ed = Editor::with_text(self.new_tab_ctx(rpc_peer), view_id, contents);
+                self.buffers.add_editor(view_id, &buffer_id, ed, Some(path));
             }
             Err(err) => {
-                // TODO: we should be reporting errors to the client
-                // (if this is even an error? we treat opening a non-existent file
-                // as a new buffer, but set the editor's path)
-                print_err!("unable to read file: {}, error: {:?}", buffer_id, err);
-               self.new_empty_view(rpc_peer, view_id, buffer_id);
+                let ed = Editor::new(self.new_tab_ctx(rpc_peer), view_id);
+                if path.exists() {
+                    // if this is a read error of an actual file, we don't set path
+                    // TODO: we should be reporting errors to the client
+                    print_err!("unable to read file: {}, error: {:?}", buffer_id, err);
+                    self.buffers.add_editor(view_id, &buffer_id, ed, None);
+                } else {
+                    // if a path that doesn't exist, create a new empty buffer + set path
+                    self.buffers.add_editor(view_id, &buffer_id, ed, Some(path));
+                }
             }
-        }
+        }; 
```

Comment:
(I don't want to be a nitpicker, but since I am adding comments anyway ... semicolon could be omitted for consistency)